### PR TITLE
[SPARK-47099][SQL] Use `ordinalNumber` to uniformly set the value of `paramIndex` for the error class `UNEXPECTED_INPUT_TYPE`

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to use `ordinalNumber` to `uniformly` set the value of `paramIndex` for the error class `UNEXPECTED_INPUT_TYPE`.

### Why are the changes needed?
When I was reviewing the spark code, I found that:
- Some expressions may have a starting value of 1 when throwing error `DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE`, eg:
https://github.com/apache/spark/blob/b0aad59f123581b66515c864873f46ea4ec4e762/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L300-L302

- Some are 0, eg:
https://github.com/apache/spark/blob/b0aad59f123581b66515c864873f46ea4ec4e762/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitmapExpressions.scala#L117-L119

**We should `unify` it and avoid understanding differences.**

### Does this PR introduce _any_ user-facing change?
Yes, the value of 'paramIndex' for the error class `UNEXPECTED-INPUT-TYPE` is uniformly set by `ordinalNumber`.

### How was this patch tested?
- Updated existed UT.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @cloud-fan @MaxGekk

**Review by MaxGekk — COMMENTED** [review]
> We should unify it and avoid understanding differences.

I agree we should unify. BTW, did we document somewhere that param indexes are 1-based?

**Comment by cloud-fan** [comment]
Shall we make the error message more explicit?
```
The <paramIndex>th parameter requires the <requiredType> type, ...
```

or something more accurate in English. cc @srielau

**Comment by panbingkun** [comment]
> > We should unify it and avoid understanding differences.
> 
> I agree we should unify. BTW, did we document somewhere that param indexes are 1-based?

How about this file?
https://github.com/apache/spark/blob/f0f35c8b1c8f3b1d7f7c2b79945eb29ffb3c8f9a/common/utils/src/main/resources/error/README.md?plain=1

Add the following content at the end of this file?
```
## Extra
1.The parameter paramIndex of error class  `DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE` is 1-based.
```

**Comment by panbingkun** [comment]
> Shall we make the error message more explicit?
> 
> ```
> The <paramIndex>th parameter requires the <requiredType> type, ...
> ```
> 
> or something more accurate in English. cc @srielau

How about the following error message?
```
The <paramIndex>-th parameter of function <functionName> requires the <requiredType> type, however <inputSql> has the type <inputType>.
```

I found a reference:
https://github.com/alibaba/sentinel-golang/blob/08071855bc67423777353c6e02f1390b419172b1/core/hotspot/rule.go#L79-L80
image

**Comment by MaxGekk** [comment]
This `<paramIndex>th` seems oddly: `1th` instead of `1st`, `2th` instead of `2nd`.

How about to reuse the function:
https://github.com/apache/spark/blob/fb1e7872a3e64eab6127f9c2b3ffa42b63162f6c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L570-L574
in the error message.

**Comment by panbingkun** [comment]
> This `<paramIndex>th` seems oddly: `1th` instead of `1st`, `2th` instead of `2nd`.
> 
> How about to reuse the function:
> 
> https://github.com/apache/spark/blob/fb1e7872a3e64eab6127f9c2b3ffa42b63162f6c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L570-L574
> 
> 
> in the error message.

Okay, I think it's great, let me do it.

**Comment by panbingkun** [comment]
Considering that we have already used `ordinalNumber` to `uniformly` set the value of `paramIndex` for the error class `UNEXPECTED_INPUT_TYPE`, it is more reasonable for us to use `0-based` here.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/BitmapExpressionsQuerySuite.scala [hunk 1]
```diff
@@ -217,7 +217,7 @@ class BitmapExpressionsQuerySuite extends QueryTest with SharedSparkSession {
       errorClass = "DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE",
       parameters = Map(
         "sqlExpr" -> "\"bitmap_count(a)\"",
-        "paramIndex" -> "0",
+        "paramIndex" -> "1",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/BitmapExpressionsQuerySuite.scala]: Should be `first`?

File: sql/core/src/test/scala/org/apache/spark/sql/BitmapExpressionsQuerySuite.scala [hunk 2]
```diff
@@ -239,7 +239,7 @@ class BitmapExpressionsQuerySuite extends QueryTest with SharedSparkSession {
       errorClass = "DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE",
       parameters = Map(
         "sqlExpr" -> "\"bitmap_or_agg(a)\"",
-        "paramIndex" -> "0",
+        "paramIndex" -> "1",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/BitmapExpressionsQuerySuite.scala]: ditto

**Review by MaxGekk — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/BitmapExpressionsQuerySuite.scala (hunk 1 above)
Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/BitmapExpressionsQuerySuite.scala]: Yea, `two` were `missed` and have been fixed.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @cloud-fan for review.

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @panbingkun and all.

This seems to break Java 21 environment too.
- https://github.com/apache/spark/actions/runs/8020316098/job/21917446965

This PR missed the following while updating `try_arithmetic.sql.out`.
- sql/core/src/test/resources/sql-tests/results/try_arithmetic.sql.out.java21

Here is a follow-up.
- #45235

**Pull request closed without merging**